[mlir][affine] Add `AffineScope` trait to `scf.index_switch`

gptsarthak opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/64287

The code in AffineStructures.cpp:FlatAffineValueConstraints' assert statement checks that the loop bound (i.e. val) isTopLevelValue or isForInductionVar. In this case, `%dim = memref.dim %20, %c0 : memref<?x?x6xi1>` is neither, since it is inside `scf.index_switch`, which is not a part of `AffineScope`. Thus, the assert is triggered.

However, `%dim = memref.dim %20, %c0 : memref<?x?x6xi1>` is an invariant in the scope of the `scf.index_switch`. If it was a loop,  a invariant code motion could have hoisted it to be outside the `scf` scope, making it a valid `TopLevelValue`.

Adding the `AffineScope` trait to `scf.index_switch` solves the issue. This defines a new top level scope for symbols, which proves highly practical as it enables a broader range of things to be represented as sequences of affine loop nests.
I even suggest that we add this trait to other SCF operations like `scf.for`, `scf.while` as well, as that will be very useful, and should not have any major issues that we cannot solve.

bondhugula commented on mlir/test/Dialect/Affine/parallelize.mlir:
```diff
@@ -323,3 +323,22 @@ func.func @iter_arg_memrefs(%in: memref<10xf32>) {
+  default {
+    %dim = memref.dim %1, %c0 : memref<?xi1>
+    affine.for %arg0 = 0 to %dim {
+            %3 = affine.load %1[%arg0] : memref<?xi1>
+            affine.store %3, %alloc[%arg0] : memref<?xi1>
```
Indent with two spaces.

gptsarthak replied:
Fixed. Thank you for reviewing my first patch in affine!

bondhugula reviewed (approved):
LGTM

gptsarthak committed:
Add `AffineScope` trait to `scf.index_switch`

Adding the `AffineScope` trait to `scf.index_switch` solves https://github.com/llvm/llvm-project/issues/64287. This defines a new top level scope for symbols, which proves highly practical as it enables a broader range of things to be represented as sequences of affine loop nests.
I suggest that we add this trait to other SCF operations like `scf.for`, `scf.while` as well, as that will be very useful, and should not have any major issues that we cannot solve.

bondhugula commented:
This isn't actually the right fix to https://github.com/llvm/llvm-project/issues/64287 - the analysis methods needed to properly return status. Fixed here: https://github.com/llvm/llvm-project/pull/129476
